### PR TITLE
Interdimensional compatibility

### PR DESCRIPTION
**Pull request by rowleya**
Makes it possible for vertices with different dimensions to communicate correctly.  This means that any connectors can be used between vertices and they should still work.

Depends on:
SpiNNakerManchester/PACMAN#516
SpiNNakerManchester/SpiNNFrontEndCommon#1064
SpiNNakerManchester/PyNN8Examples#104

### CODE REVIEW

File: unittests/mocks.py [hunk 1]
```diff
@@ -34,3 +37,13 @@ def label(self):
 
     def __repr__(self):
         return "Population {}".format(self._label)
+
+    @property
+    def _vertex(self):
+        return self._mock_vertex
+
+
+class MockVertex(object):
```
Comment (Christian-B) [inline unittests/mocks.py:46]: Why not use pacman_test_objects/simple_test_vertex.py and extend that with the extra method

optional!

File: unittests/model_tests/neuron/test_synaptic_manager.py [hunk 2]
```diff
@@ -113,13 +113,14 @@ def test_write_data_spec():
         additional_parameters={
             "splitter": SplitterAbstractPopulationVertexFixed()})
     proj_one_to_one_1 = p.Projection(
-        pre_pop, post_pop, p.OneToOneConnector(),
+        pre_pop, post_pop, p.FromListConnector([(i, i) for i in range(10)]),
```
Comment (Christian-B) [inline unittests/model_tests/neuron/test_synaptic_manager.py:116]: Why remove the check of a OneToOne and an AllToAll?

If wanted ADD a fromList or convert one of the OneToOne

File: unittests/test_using_virtual_board/test_from_file_connector.py [hunk 3]
```diff
@@ -90,16 +90,6 @@ def test_simple(self):
         ]
         self.check_other_connect(as_list)
 
-    def test_list_too_big(self):
```
Comment (Christian-B) [inline unittests/test_using_virtual_board/test_from_file_connector.py:93]: Why remove this test?

If it no longer works check the pop

File: unittests/test_using_virtual_board/test_from_list_connector.py [hunk 4]
```diff
@@ -78,16 +78,6 @@ def test_simple(self):
         ]
         self.check_other_connect(as_list)
 
-    def test_list_too_big(self):
-        as_list = [
```
Comment (Christian-B) [inline unittests/test_using_virtual_board/test_from_list_connector.py:82]: Either it work and the test stays or it does not and we check for a good error

File: spynnaker/pyNN/models/neuron/population_machine_neurons.py [hunk 5]
```diff
@@ -335,10 +319,16 @@ def _write_current_source_parameters(
         spec.switch_write_focus(self._neuron_regions.current_source_params)
 
         # Get the current sources from the app vertex
+        app_current_sources = self._pop_vertex.current_sources
         current_source_id_list = self._pop_vertex.current_source_id_list
 
         # Work out which current sources are on this core
-        current_sources = self.__get_current_sources_sorted()
+        current_sources = set()
```
[the remark this answers is not in the document]
Comment (rowleya) [reply spynnaker/pyNN/models/neuron/population_machine_neurons.py]: Changed to OrderedSet

File: spynnaker/pyNN/models/projection.py [hunk 6]
```diff
@@ -525,3 +528,7 @@ def size(self, gather=True):  # @UnusedVariable
         """
         # TODO
         _we_dont_do_this_now()
+
```
[the remark this answers is not in the document]
Comment (rowleya) [reply spynnaker/pyNN/models/projection.py]: Removed and fixed the test

File: spynnaker_integration_tests/test_various/test_synaptic_expander.py [hunk 7]
```diff
@@ -48,18 +48,15 @@ def run_script():
     connectors = [
         (p.OneToOneConnector, functools.partial(check_one_to_one, 10)),
         (p.AllToAllConnector,
-         functools.partial(check_all_to_all, 10, True)),
-        (functools.partial(p.AllToAllConnector,
```
[the remark this answers is not in the document]
Comment (rowleya) [reply spynnaker_integration_tests/test_various/test_synaptic_expander.py:52]: This test is done elsewhere now as the current test doesn't do self-connection (so the test fails its own check)

File: unittests/connector_tests/test_csa_connector.py [hunk 8]
```diff
@@ -1,111 +0,0 @@
-﻿# Copyright (c) 2017 The University of Manchester
```
[the remark this answers is not in the document]
Comment (rowleya) [reply unittests/connector_tests/test_csa_connector.py:1]: This is done as an integration test now, as it wasn't able to pass as a unit test.

File: unittests/mocks.py (hunk 1 above)
Comment (rowleya) [reply unittests/mocks.py:46]: Could have... but then I was changing sPyNNaker here!

File: unittests/model_tests/neuron/test_synaptic_manager.py (hunk 2 above)
Comment (rowleya) [reply unittests/model_tests/neuron/test_synaptic_manager.py:116]: Undone changes - must have been testing something here!

File: unittests/test_using_virtual_board/test_from_file_connector.py (hunk 3 above)
Comment (rowleya) [reply unittests/test_using_virtual_board/test_from_file_connector.py]: Restored - must have removed thinking it couldn't work in unit test, but actually found issues!

File: unittests/test_using_virtual_board/test_from_list_connector.py (hunk 4 above)
Comment (rowleya) [reply unittests/test_using_virtual_board/test_from_list_connector.py]: Restored as above!

File: spynnaker/pyNN/models/neuron/population_machine_neurons.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/neuron/population_machine_neurons.py]: Actually changed to list (by sorted method) which is the correct fix

**Review by Christian-B — APPROVED** [review]
Please link all related branches so we don't miss one

**Pull request merged**